community/tutorials/gdnative/gdnative-c-example.rst

```diff
@@ -0,0 +1,328 @@
+
+.. code:: C
+
+    typedef struct user_data_struct {
+        char data[256];
+    } user_data_struct;
+
+And then we define our constructor. All we do in our constructor is allocate memory for our structure and fill it with some data. Note that we use godots memory functions so the memory gets tracked and then return the pointer to our new structure. This pointer will act as our instance identifier in case multiple objects are instantiated.
+
+This pointer will be passed to any of our functions related to our object as a parameter called p_user_data and can both be used to identify our instance and to access its member data.
+
+.. code:: C
+
+    void *simple_constructor(godot_object *p_instance, void *p_method_data) {        
+        user_data_struct *user_data = api->godot_alloc(sizeof(user_data_struct));
+        strcpy(user_data->data, "World from GDNative!");
+
+        return user_data;
+    }
+
+Our destructor is called when Godot is done with our object and we free our instances member data.
+
+.. code:: C
+
+    void simple_destructor(godot_object *p_instance, void *p_method_data, void *p_user_data) {
+        api->godot_free(p_user_data);
+    }
+
+And finally we implement our get_data function. Data is always sent and returned as variants so in order to return our data, which is a string, we first need to convert our C string to a Godot string object, and then copy that string object into the variant we are returning.
+
+Strings are heap-allocated in Godot, so they have a destructor which frees the memory. Destructors are named godot_TYPENAME_destroy. When a Variant gets created with a String it references the String. That means that the original String can be "destroyed" to decrease the ref-count. If that does not happen the String memory will leak since the ref-count will never be zero and the memory never deallocated. The returned variant gets automatically destroyed by Godot.
+
+(In more complex operations it can be confusing the keep track of which value needs to be deallocated and which does not. As a general rule: call godot_XXX_destroy when a C++ destructor would be called instead. The String destructor would be called in C++ after the Variant was created, so the same is necessary in C)
+
+The variant we return is destroyed automatically by Godot.
+
+.. code:: C
+
+    godot_variant simple_get_data(godot_object *p_instance, void *p_method_data
+            , void *p_user_data, int p_num_args, godot_variant **p_args) {
+        godot_string data;
+        godot_variant ret;
+        user_data_struct * user_data = (user_data_struct *) p_user_data;
+
+        api->godot_string_new_data(&data, user_data->data, strlen(user_data->data));
+        api->godot_variant_new_string(&ret, &data);
+        api->godot_string_destory(&data);
+
+        return ret;
+    }
+
+And that is the whole source code of our module.
```

Comment:
If you add a blank `` .gdignore`` file to the src folder, Godot will not try to import the compiler-generated temporary files.